File: 1493-longest-subarray-of-1s-after-deleting-one-element/1493-longest-subarray-of-1s-after-deleting-one-element.java
class Solution {
    public int longestSubarray(int[] nums) {
        // [0,1,1,1,0,1,1,0,1]
        //    j         r
        //  c - 1
        int left = 0;
        int count = 0;
        int length = 0;
        if(nums[left] == 0) {
            count++;
        }
        
        for(int right = 1; right < nums.length; right++) {
            if(nums[right] == 0) {
                count++;
            }
            
            while(count > 1) {
                if(nums[left] == 0) {
                    count--;
                }
                left++;
            }
            
            length = Math.max((right - left), length);
           
                        
            
        }
        
        return length;
    }
}

File: 1493-longest-subarray-of-1s-after-deleting-one-element/1493-longest-subarray-of-1s-after-deleting-one-element.py
"""
sliding window problem
- keep expanding the window when you come across a one
- keep track of the number of zeros
- while num_zeros > 1 shrink the window
- keep track of max_length
    - return max_length - 1
    
 0 1 1 1 0 1 1 0 1
           s     f

zeros = 1
max_len = 6
"""

class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        slow = 0
        num_zeros = 0
        max_len = 0
        
        for fast in range(len(nums)):
            if nums[fast] == 0:
                num_zeros += 1
            while num_zeros > 1:
                if nums[slow] == 0:
                    num_zeros -= 1
                slow += 1
            max_len = max(max_len, fast - slow + 1)
            
        return max_len - 1